capstone/src/main/webapp/js/autocomplete.js

```diff
@@ -0,0 +1,148 @@
+const SEARCH_INPUT = document.getElementById('search');
+const ITEM_CLASS = 'autocomplete-items';
+const ACTIVE_CLASS = 'autocomplete-active';
+const LIST_ID = 'autocomplete-list';
+const UP_KEY = 40;
+const DOWN_KEY = 38;
+const ENTER = 13;
+let selectedElmntPos = -1;
+
+/**
+ * Adds the 'active' tag to an autocomplete elmt
+ * @param {Object} autocompleteListElmt
+ * @return {bool}
+ */
+function addActiveTag(autocompleteListElmt) {
+  if (!autocompleteListElmt) {
+    return false;
+  }
+  removeActiveTag(autocompleteListElmt);
+  if (selectedElmntPos >= autocompleteListElmt.length) {
+    selectedElmntPos = 0;
+  } else if (selectedElmntPos < 0) {
+    selectedElmntPos = (autocompleteListElmt.length - 1);
+  }
+  autocompleteListElmt[selectedElmntPos].classList.add(ACTIVE_CLASS);
+}
+
+/**
+ * Removes the 'active' tag from an autocomplete elmt
+ * @param {Object} autocompleteListElmt
+ */
+function removeActiveTag(autocompleteListElmt) {
+  for (elmt of autocompleteListElmt) {
+    elmt.classList.remove(ACTIVE_CLASS);
+  }
+}
+
+/**
+ * Closes dropdown autocomplete list
+ * Does not close the element provided
+ * @param {Object} elmnt
+ * @return {void}
+ */
+function closeAllElmntExcept(elmnt) {
+  const autoItems = document.getElementsByClassName(ITEM_CLASS);
+  for (const item of autoItems) {
+    if (item != elmnt && item != SEARCH_INPUT) {
+      item.parentNode.removeChild(item);
+    }
+  }
+}
+
+/**
+ * Current placeholder until servlet is created
+ * @return{Array<string>} arr
+ */
+function getListOfSuggestions() {
+  const arr = ['Hello', 'Hi', 'Howdy'];
+  return arr;
+}
+
+/**
+ * Identifies and returns relevant suggestions in the arr
+ * @param {string[]} arr
+ * @param {string} val
+ * @return {string[]} trimArr
+ */
+function getRelevantDataSuggestions(arr, val) {
+  const trimArr = [];
+  val = val.toUpperCase();
+  for (const arrElt of arr) {
+    if (arrElt.substr(0, val.length).toUpperCase() == val.toUpperCase()) {
+      trimArr.push(arrElt);
+    }
+  }
+  return trimArr;
+}
+
+/**
+ * Takes relevant suggestions and displays them in DOM
+ * @param {Object} trimArr
+ * @param {Object} autocompleteList
+ * @param {string} val
+ * @return {void}
+ */
+function displaySuggestions(trimArr, autocompleteList, val) {
+  for (arrElt of trimArr) {
+    const listElmt = document.createElement('DIV');
+    listElmt.innerHTML = '<strong>' +
+                         arrElt.substr(0, val.length) +
+                         '</strong>';
+    listElmt.innerHTML += arrElt.substr(val.length);
+    listElmt.addEventListener('click', (e) => {
+      SEARCH_INPUT.value = arrElt;
+      closeAllElmntExcept();
+    });
+    autocompleteList.appendChild(listElmt);
+  }
+}
+
+/**
+ * Event occurrance when input is provided to autocomplete field
+ * @return {void}
+ */
+function giveSuggestions() {
+  closeAllElmntExcept();
+  const val = SEARCH_INPUT.value;
+  if (!val) {
+    return false;
+  }
+  const arr = getListOfSuggestions();
```

Comment:
So i was thinking about the async problem @ccondit brought up in the standup. One solution that came to my mind was to check if input data is still the same value after the fetch completed. Basically right here you can check if `SEARCH_INPUT.value == val`. IF this check failed, it means there is a new query and the suggestions for the old one can be discarded right here. Just a suggestion. I'm not sure how it would that affect the overall auto-complete flow.

Reply:
I tried this approach kaveh and it doesn't seem to work :/ I am thinking of ways to address this problem in the following PR (part 2)

Reply:
added the suggestion btw and it works fine now :) it'll be telling to see how it updates in the next pr

Reply:
Trying to grok this with [this comment](https://github.com/googleinterns/univiz/pull/50#discussion_r478592139):

@mmurray22 - looks like it did get removed?

also look like @kaveh096 is suggesting adding this *after* the fetch call is completed, not before...